Enemy DS Logic Update

r0bd0g opened the pull request:
Sorry there's so much text here. Every time I made an adjustment that actually affects something I made a note about it, and I'm pasting all of those notes here now.

Other Changes:
- General 'moving stuff around for no real reason' changes throughout. If I didn't call something out anywhere here, then it's not (supposed to be) a practical change (or I forgot to write it down lol). Ask about anything you don't see documented here.
- Brought in some of the new logic changes from main branch, including the logic changes from Ganon's Tower ER (not the non-logic-related changes in the world files). I wanted to have them in just in case there was something about them that needed handling for the enemy/souls PR. I also brought in the changes for the Jabu MQ Explosive Wonderitems to this branch since TBH wonderitems were a bit of a side focus in this review, being a relatively new feature added since the last full review.
- [Applies to main branch] Beehives with Bombchus trick was tagged as Fountain rather than Domain.
- [Applies to main branch] Renamed preexisting instances of "Redead" to "ReDead". I don't know if this will break any plandos or not. I suspect not. There was also one instance of "like-like" changed to "Like Like".
- Renamed some locations for consistency with existing checks in the area, or with their MQ/Vanilla counterparts, or renamed them if the given enemy name was wrong, or renamed some to include more detail about the specific check rather than just a number.
- [Applies to main branch] Adjustments to locations tags, and handful of which apply to main branch.
- [Applies to main branch] Corrected the wonderitem vanilla drops. (I decided to go with Blue Rupee for wonderitems that normally give random drops.)
- Added "kwargs" to region_has_shortcuts, or something?

Logic Helpers:
(When I say "Applies to main branch" in this section, I just mean I have adjusted what was considered acceptable to kill that type of enemy compared to what we previously considered acceptable on main branch.)
- Octorok: Added Hookshot. Reworked the "with_explosives" helper to "with_bombs". It came up too often so I had to add a "with_bombs_or_magic" helper. Also added a with_explosives helper that actually checks for chus, but only in combination with magic b/c without magic never came up.
- Scrub: Added Nuts+Explosives.
- Peahat: Added bombchus for killing them. At day, you can use chus to kill them without waking them up and having to fight them. It's pretty annoying to get the right timing for the bombchus, but as long as the day holds out you're pretty safe and it's honestly sometimes preferable to actually fighting them. Added Din's, which one-shots them out of the ground at night. I don't think the Deku Shield is really needed when using Bombs -- just run in a couple circles to dodge the babies. The peahat logic was broken up into day/night even though they can only be fought in areas where time passes. I went ahead and just simplified the logic with the assumption that they would only be fought at their current time-passing locations. (GS's in time passing areas usually check at_night but I think being so precise with Peahats was adding a lot of unnecessary complication.)
- Flying Peahat: Added Din's Fire, and removed Deku Shield for when using Bombs.
- Guay: Added a helper for them. Removed explosives from the possible requirements since this enemy is annoying enough already, especially at Colossus.
- Stalfos: Removed sticks from the helper.
- Biri/Bari: I added the Bari helper. I definitely changed the Biri helper, but I didn't write down the specifics. Baris differ from Biris in that boomerang stuns instead of killing the Baris.
- Like Like: Removed bombchus, added slingshot.
- Shabom: Added roll invincibility. For the record, if this is too hard, the logic becomes "has_soul(Shabom) and (is_adult or can_take_damage or Nuts or Kokiri_Sword or Sticks or Boomerang or can_use(Dins_Fire))". In the entrance of vanilla Jabu, you'd have to also consider bringing Ruto from the main rooms.
- Tailsparan: Added Hammer.
- [Applies to main branch] Lizalfos: Removed bombs without nuts. Added with_bombs helper for special cases where you can chuck bombs at them from safety.
- Dinolfos: Removed Bombs. I'm over it. Also removed Slingshot without nuts, I'm over that too. They're too elusive.
- [Applies to main branch] Keese: Removed explosives without nuts. Added "stationary" helper for when explosives can still be used without nuts.
- Skulltula: Removed boomerang (doesn't kill them).
- [Applies to main branch] Iron Knuckle: Added bombchus. Removed sticks without Slingshot.
- Flare Dancer: Added a helper (probably won't get a lot of use though).
- Torch Slug: Added a helper.
- [Applies to main branch] Freezard: Added helper. Added sticks. (You have to jumpslash. I'm not sure if the reason we didn't ask for sticks before is because they're consumable, and regular slashes look like they do damage? And also some of their more prominent rooms feature a lot of Freezards, so it'd take a lot of sticks to kill them all.)
- [Applies to main branch] Removed Bubble helper. Added Blue Bubble and Green Bubble helpers. Blue Bubbles need to be stunned first, so there's a complex series of requirements to do that. Removed bombchus from blue bubbles. Added child Hylian Shield for stunning them (tilt it up to make it more likely to hit). Green Bubbles drop their shields occasionally, so no stunning required. (I'm not sure this green bubble helper is getting a lot of use though.)
- Walltula: Removed helper. A few of them could maybe have used a basic helper of is_adult or can_child_attack but I felt that b/c the vast majority of them needed logic specific for each one it'd be more consistent to just not use a helper.
- Floormaster: Removed Bombs without nuts.

Overworld
Kokiri:
- Added explosives (bombs was missing, but bombchus are also fine here since it's an easy straight shot and they usually will be low enough to automatically explode the chu) or Magic as options for the Octoroks.
- [Applies to main branch] I removed Deku Baba Nuts/Sticks as adult as the various settings that would or wouldn't allow obtaining these drops to be in logic were wrong and fixing it was getting too convoluted.
- But the current way that the enemies in Kokiri Forest work (though I'm not sure they should work quite like this) makes it so the nut/sticks from the Babas really are always there with no ER jank to worry about, so we'll bring them back only under the new setting.
- [Applies to main branch (except mixed pools isn't implemented yet)] I noticed that gossip stone fairy was missing near Deku Tree, which could be relevant if you were to come into child Kokiri from Deku without the ability to pass Mido to reach the existing gossip stone in main Kokiri.
Lake:
- Added bombs for the child Octoroks.
Valley:
- Moved the Octoroks into the correct regions.
- [Applies to main branch] And this is how I noticed the valley wonderitems were in the wrong region as well.
- [Applies to main branch] I decided actually probably the bean should be the default logic for these valley wonderitems. They're actually still pretty annoying without bean even with the visual indicator. New trick added to get them without the bean.
- [Applies to main branch] Reversed the names of the valley wonderitems, since I heard that they were backwards.
- Added Hylian Shield+Irons to lowest Octorock.
Graveyard:
- The ReDeads in Dampe's Grave cannot be killed with Din's Fire because Din's Fire is disabled in that area.
Trail:
- Adult Walltulas: Jumpslash is safe against all of them.
- Child Lower/Middle Walltulas: can_child_attack works fine. (Highest one with sword items too awkward for default.)
Crater:
- I think the three non-first plank DMC Bubbles nearest the Bolero bad are all safe when using magic under FTR and coming in from Bolero.
- Added the first plank bubble. You have to throw a Bomb at where the Bubble will land, and then use the Hover Boots to get low enough to spawn it without having to fall into the lava, and then hopefully it'll jump into the bomb explosion, and then you have to do all of that again. It's probably too hard for default logic but if it's the only realistic way to do this check, then I guess it's what you gotta do to do the check. And it seems a shame not to shuffle it when everything else in the game is shuffled. We might have to figure something out so that people are less likely to get caught by this check unawares.
River:
- Add magic to all the adult Octoroks. Added Bombchus all but the lowest two octoroks.
Fountain:
- The two Octoroks nearest the circling platforms can't realistically be reached with bombs.
- New lens trick for the skulltulas in the hidden cave. Note that the one at the top isn't invisible. I added it to the hell mode preset. I named it after Fountain since it only applies there, but the internal name is "overworld" so that the trick can be easily renamed if more invisible OW things turn up later.
- Added commented out logic for the Wonderitem that doesn't exist, not sure why honestly.
Castle:
- Walltulas in the storms grotto are behind a bombable wall so, let's require that blown up, since it's opaque. Except if you have Din's then I guess you don't need to know exactly where the Walltula is to kill it, so we'll allow Din's.

Bosses
- Adjustments to Keese and Dinolfos stemming from changes to their respective helpers.

MQ Deku
- Lobby Walltula safe with sword items.
- Torches room had not been properly connected up with the rest of the rooms.
- The Basement skulltula can be killed from across the room with Bow or Bombchus.
- [Applies to main branch] MQ Deku Water Back to Larva Room with Sticks (I think? it's confusing lol) needs to confirm you have a way to cross the room as child, so I added a check for that.

MQ DC
- Add lower staircase Beamos with strength.
- Moved the staircase room scrub from lower to upper staircase region.
- SoT Skulltulas: can_child_attack is being used but it needs to be can_child_damage b/c Boomerang doesn't work. Bombchus can be used without Song of Time. It's a straight, visible shot that automatically explosives the bombchus.
- Require SoT for the Walltula. Allow Hammer or Din's as alternatives.
- Full logic for blowing up the staircase needed. Trick+Bow was missing.
- Revert adding can_break_crate to larvae room GS (guaranteed to be able to kill it through the create, and we don't care about crate opaqueness in the context of bonko).
- That skulltula that is closest to the webs I moved to the previous region, since you can kill it through the webs.
- Added weapon for keese and baby dodongos in before upper lizalfos room. We could be in here with strength and din's as child, which doesn't kill these enemies. (There are pots but I didn't think using those would be fair on these enemy types.)
- Except one of the keese is on top of a bomb flower and can be killed with Din's.
- Before upper lizalfos exit was bypassing the new upper lizalfos region, going directly into torch puzzle upper, erroneously allowing you to skip the Lizalfos soul.
- Add a weapon check for the keese after the Lizalfos that's on the bomb flower; he's kinda rude. The other keese is fair with pot or bomb flower.
- You can use a pot on the lower right side skulltula. Pots/Flowers/Din's all seemed a bit too much for use on that mobile keese. I don't know if any of this matters because I feel like a weapon is probably guaranteed to even get to lower right side.
- Added Bow to go from lower right side to lizalfos room, so Adult can kill the baby dodongos without needing the lizalfos soul.
- The lower lizalfos don't spawn if you fall down from above, but they were in the region that you fall down from above into.
- Reallow bombs for lower lizalfos (you can camp from the secret alcove).
- [Applies to main branch] The lower lizalfos hidden freestanding was sharing a name with the vanilla variant, but this is not how checks in scenes that differ between the MQ and Vanilla are handled. I guess the functionality that would be lost is the ability to disable the check conditionally based on whether the dungeon is MQ or Vanilla.
- We're not using pots on baby dodongos, so let's check for weapons in the near boss areas.
- But we are using bomb flowers on armos and lizalfos back here. Also continue to allow Bombs on this Lizalfos since you can easily camp from the higher ground.

MQ Jabu
- Lobby Octorok and Shaboms with pots.
- Basement Biri and Shabom with Ruto.
- Boomerang chest room Stingers with pots.
- Reallow Bombs on the Boomerang chest Lizalfos. You can camp from down in the water, or if he falls into the water, you can camp from above the water.
- Use the new Bari helper for Baris.
- [Applies to main branch] Added MQ Jabu Boomerang chest room SoT GS kill with pot.
- [Applies to main branch] Added the bombable wonderitems as a logical Fairy source, since it is repeatable.
- The logic for the invisible keese seemed to be based on the logic for only the farthest away keese. But even that one could also have been reached with Hover Boots. Besides Hovers, Hookshot and Boomerang will reach the left one. The right one is reasonable stir up (sorry about the Stingers) and then kill up close as you would a normal keese (I mean except that these ones are invisible).
- Invisible enemies GS shouldn't have been always checking for the keese and stinger souls since there are some route to the GS that don't need to kill those enemies.

MQ Botw
- Add lens for the invisible enemies. Add new lens trick for MQ BotW.
- Center room Wallmaster with pots.
- Add explosives or sticks to hitting the switch after the crawlspace.
- Add Din's Fire from the main room to the Walltulas, through the gate.

MQ Forest
- Add jumpslash and fall damage, or bombs, to all but the highest Walltula. Except the one nearest the top of the tree can be done without fall damage.
- Removed "Wolfos Clear" event, which was forcing SoT/Child access when it shouldn't. Also allow dropping down from above to kill the Wolfos because it shouldn't be possible to kill them and avoid getting their drops, right?
- Added Floormaster soul to 2-key route for after block puzzle to lower west courtyard access.
- Added logic to hit the switch to access the straightened hallway. Had to rework the regions a little bit to make this work. The typical route would be with floormaster soul, but there are trick routes to hit the switch early.
- The Floormaster can be accessed by falling down from above, sure, but there are trick routes to the outside upper ledge early, and so you could access the Floormaster from behind.
- Added adult+magic to the Octoroks.
- Added master sword, explosives, Boomerang, or Din's to lowest courtyard Walltula. Added bombchus, hookshot, or boomerang to the middle courtyard Walltula. Added Longshot or Din's (Din's only if you reached the upper balconies) to the highest courtyard Walltula.
- The compass chest wasn't checking for the Wolfos soul.

MQ Fire
- Map Chest was in the wrong region. It needs to be after the Iron Knuckle, (Spinning Tiles?,) and Flaredancer souls.
- The Walltula wasn't checking for its soul. Let's kill it through the wall only with Hammer or Din's, since with those you don't need to know exactly where the Walltula is.
- Torch slugs were either in the wrong region or weren't duplicating the full logic to get to that region.
- Two wonderitems after the Flare Dancer were not checking for the soul.
- Removed Stalfos soul from top GS.

MQ Water
- Stalfos soul not necessary for the map chest. The hookshot is what spawns the chest.
- Hookshot not necessary for the map chest Stalfos. The hookshot was just to spawn the chest.
- Stalfos soul is not needed for before Dark Link lower pots. You can hookshot over to them.
- Neither Stalfos soul nor Dark Link soul are needed for the before Dark Link lower pots' Fairy and Nut drops. You can hookshot over to them.
- You need Hovers or Scarecrow to get over to the triple wall torch Stalfos.
- If you're using the trick to reverse through the Dodongo room, you could have been expected to reach the 3 spout wonder items without having the Dodongo soul. New region added for those water spout items.
- [Applies to main branch] It used to be that we didn't need to confirm that you could raise the water level back up if you were to lower it, because everything that required the highest level required hookshot anyway. But that assumption is broken by the map chest Stalfos... But actually it was broken a while ago with the crates at the top of central pillar, and the crates at the bottom of the central pillar when using Din's, but we never noticed.
- [Applies to main branch] A while back I realized also that you could lock yourself out of the lowest water level by bringing it up to the middle. In all of these cases you need to be able to restore the water back to the top if it were to be lowered, which requires hookshot or hovers. I added a new event to MQ for this purpose. I guess this change is mildly related to drops/souls since a handful of the new checks are affected, but this is mainly a main branch bug fix.

MQ Shadow
- The skulltula behind the gate can be killed with a jumpslash, no need for the rupees.
- For consistency I decided to add magic as a requirement for the huge pit bubble.
- The upper huge pit beamos can be killed with a bomb flower. Strength, bow, and Din's all work, but Din's can make the Beamos briefly immune to explosions in such an awkward way that I don't think I can require it.
- The Like Like silver rupee is not a Like Like, so we don't need the soul.
- The lower huge pit beamos had a brackets error that wasn't saved by operator precedence, but I fixed it just by moving it into the region with the rest of the checks.
- Added the redead soul for obtaining the two silver rupees that 'require' getting on top of the chest to reach without Longshot.
- But you can also skip the redead soul for those rupees and for progressing through the dungeon by already having the silver rupees, which spawns glass blocks you can use instead.
- And you can skip redead soul for those silver rupees by coming in through the upper door. Shortcuts specific logic would be needed for these checks if Longshot wasn't one of the options to get to them from the lower door into the room, since Longshot is "required" to reach the locked door at the end of the dungeon.
- You need the Hookshot to reach the wind tunnel Skulltula.
- Added the flying pot soul to the enemy chest in the after wind room (but I'm not happy about it lol).
- Added Hammer to the before boat Walltula, which can be performed from inside the shortcut area without needing full access to the boat.

MQ Spirit
- [Applies to main branch] The silver block hallway chest was in the child region and wasn't implementing the logic to hammer the boulder through the crawlspace to reach shared Spirit. I just went ahead and moved it into shared spirit despite being child-only.
- [Applies to main branch] I noticed that the hammer through walls trick route didn't have a way to hit the crystal switch when passing through to desert colossus as child. Everything else that you might want to do that's child-only requires a weapon, that you could then use on the switch. So I had to add logic for hitting it. There's a pot but it's impossible to throw it at the switch as far as I could tell, but you can shield drop it onto it.
- The keese in the map room all have different requirements. The one on the left you're guaranteed to have a weapon by killing the slugs in the previous room. The others can be reached with slingshot or boomerang as currently prescribed, as well as stirred up with Din's Fire for a direct kill. Nuts can still all of them up as well, but the one on the torch seems to always return to his perch. The lower two in the pit can be killed with bomb throws. The one below the gate can be killed with a straight-forward bombchu use (which stops working if the gate is down, but slingshot is required to lower said gate).
- Added Din's Fire to the child Gibdos.
- [Applies to main branch] I decided to be a bit more precise with the logic around this child loop. So now you have to have a way to defeat or stun the Gibdo before pulling the grave to get to the switch.
- [Applies to main branch] And I'm just gonna require K Sword on the Stalfos now like I normally do on all the other Stalfos, despite the fact that the Stalfos has a death pit. None of the options to get him into the death pit are actually all that reliable, even though you can usually count on him to fall down before too long.
- The fairy pot was in the wrong region. It's actually in the Stalfos room. You can open the pot from a distance with the slingshot you definitely have.
- The items in the pots in the Stalfos fight can all be obtained with Boomerang. For all of the pot-related things in this room, I required the Kokiri Sword, in case the Stalfos is spawned, as dealing with these pots from the spinning platform is mildly annoying enough even without distractions.
- The shared spirit Wallmaster logic was too restrictive. Technically the problem was that there was no 'shared spirit' route where you check that you have items to do it regardless of which age you are. But since either bombchus or prior adult access is guaranteed, either of which can kill the Wallmaster, the whole thing simplifies completely down.
- You can lure the adult floormaster down to kill as child.
- The bubbles were not checking for any weapon to kill them. Technically one of adult or bombchus was guaranteed, which does kill them, but good luck with that, especially with bombchus. For consistency with other bubbles I decided to require adult+magic. Bow isn't reasonable since they never respawn so you can never line it up. This unfortunately does mean there won't be any way to logically kill them as child.
- One of the Skulltulas in the symphony room is invisible, and so needs to confirm lens or trick.
- The enemies in the shifting wall room were in the wrong region. You don't need the silver rupees.

MQ Ice
- Use the helper for the Wolfos, since can_child_attack includes explosives (too annoying) or boomerang (doesn't work).
- Anything that is in logic for the Freezards can kill the Walltula.
- The two keese in the large room needed to check for a weapon to kill them. (Pots were too difficult.)

MQ GTG
- [Applies to main branch] The Dinolfos chest is supposed to be checking is_adult because it's too difficult (sometimes impossible?) for child to defeat all of these enemies within the time limit.
- One of the Dinolfos in here is actually a Lizalfos, and so using the adjusted helpers has slightly different individual kill logic.
- A new region was created for the lava room after the door, since the enemies room isn't guaranteed to be clearable anymore, so it needed to be separated out. Some of the exits around this room hadn't been properly rewired to account for this, so some routes were expecting you to kill the enemies for no reason, in order to skip bow on the lava room access.
- I moved the bubbles in the lava room into their correct regions so you could potentially spawn them in other ways than having to have access within the flames at the front of the room and maybe then needing hovers to get over to it. One of them needed a new region just for killing it. I was able to add Bow as an alternative to magic in a couple cases.
- The trick to obtain the silver rupees on the ceiling of the boulder room requires the Wallmaster.
- [Applies to main branch] The Wallmaster goes away after the silver rupees have been collected, so the trick to obtain the rupee using the Wallmaster can only be used if silver rupees are not shuffled.
- Need to check for a weapon to kill the Stalfos.
- Torch slug soul required to reach the back lava room ledge.

MQ Ganons
- More precise Shadow Trial green bubble logic.
- [Applies to main branch] Added the MQ Shadow Trial wonderitem. Sometimes the wonderitem falls off? Just reload the room I guess. The odds it doesn't fall off seemed better than 50% when I tested it.
- Loosen Forest Trial Armos logic (some methods are mildly annoying with the Beamos still there if you're not allowed to take damage, but, idk, players are gonna have to figure it out I guess).
- The Beamos next to the Armos seemed to be basically always defeated when that Armos dies.
- Added lens to the invisible keese. Stationary helper for the two stationary ones.
- Charged magic spin for the 2nd Spirit Trial Bubble.

Deku
- For the main room Walltulas, in addition to what was there before, Hammer works. Jumpslashes from above with fall damage also work. The lowest one can be killed from below with adult/ksword/sticks, but ksword no longer works if the chest is small. The middle one can be killed from below also with adult/ksword/sticks, but ksword is not free so I guess I won't call for it, and none of them continue to work if the chest is small. Except Biggoron Sword can still work from a small chest, but it's also not free so we won't call for that either (but honestly a part of the reason for not calling for either of these precise jumpslashes is that I don't want to deal with Biggoron Sword potentially not being foolish anymore).
- Back room skulltula needs to check for a way to destroy the bombable wall.

DC
- Main room Keese can be fully killed from below. They'll even fly down if you look at them longingly enough. And you can use a bomb flower which was maybe a bit rude of me to potentially require on top of the already kind of rude aspect of having to lure them down.
- We're using nuts for explosives and bomb flower kills of keese now I guess, so I added that where applicable.
- There are stationary keese in lower right side that can use the stationary logic.
- Hidden recovery heart was requiring blast/smash or lizal soul as adult due to being in the wrong region.
- Reallow bombs for lower lizalfos (you can camp from the secret alcove).
- The Walltula can be killed with a pot.
- Switch room Armos can be killed with bomb flower.
- When reversing through the upper lizalfos loop, you didn't add a reverse exit to continue to the earliest pots.
- Back room Armos wasn't checking for the bombable wall.

Jabu
- Allow also bombchus or adult+magic on the entrance Octoroks.
- Allow Ruto on main area enemies. Many of these Ruto enemy kills will only be relevant in mixed pools.
- Allow Ruto on entrance room enemies using an at() to confirm main room access.
- [Applies to main branch (except mixed pools isn't implemented yet)] Allow Ruto on near boss skulltula. It's possible we didn't include this before because in vanilla it's not normally possible to bring Ruto into this room. Need to confirm a means of carrying Ruto across the gap back to this room. The Hover Boots trick route is kind of strange because a much easier trick would be to just kill it with a jumpslash, but we don't have mixed pools specific tricks as yet.
- Allow Ruto on the rest of the near boss enemies as well.
- Wiggler platforms Octorok was in the wrong region. It needs to be after Big Octo (which confirms sword or sticks).

BotW
- None of this point practically matters since there's a guaranteed stick drop in this dungeon, but you can use pots to kill a lot of the enemies, so we'll include that in most cases just to avoid the playthrough having to pick up the stick. You can use pots on 3 of the 4 skulltulas, the Green Bubble, the Wallmaster, and the basement ReDeads.
- In the right inner room, the Keese that stirs up immediately will be using the regular Keese helper. The Keese on the wall will be using the regular stationary helper. The other two Keese are perched on nothing and so have to allow bombs but not bombchus, without using nuts. (Again none of this practically matters because of the guaranteed stick and nut drops in this area.)
- You need to blow up rocks to reach the ReDead near the map chest.
- The caged Like Like can be killed through the gate by all of the usual methods except for Slingshot.

Forest
- Fix bubble logic by using the new blue-specific helper.
- Add with_magic to the Octorok helpers.
- For the West Walltulas, I fixed the hookshot checks on all of them. I added is_adult jumpslashing it for the lower Walltula. I added Din's to the upper Walltula. I added Hammer for all of the Walltulas.
- Other methods of swimming through the well might be relevant now that the bubble might not be there, so I added in the other methods and reworked the relevant trick.
- The well recovery hearts had to be moved out of the upper balconies region now that the Iron Boots no longer guarantee access to that region.
- Added a bunch of things into logic for killing the East Walltula if you've accessed the upper balconies, mainly jumpslash plus the ability to survive fall damage. But I added a bunch of other weapons for OHKO (even though it's easy to recoil into the water if you can think to do that), namely Bombs, Hammer, Din's, and Magic (either as adult or with Kokiri Sword). Magic spin as child might not be super fair as it's only about 50/50 to work unless you know the magic only reaches if the Walltula's body is oriented downward.
- Added a region for the room below the twisted hall with the bubbles, so that you can enter that room from either direction.

Fire
- The GS in the opening loop comes before the Flare Dancer.
- One of the Big Lava Room Bubbles isn't reasonable with Bow due to not being able to spawn it from stable ground.
- Then the moving wall bubbles are kind of unreasonable with Bow due to the moving flame wall constantly disrupting you.
- The Flaredancer soul is not required for the pots before the Flaredancer.
- The Flaredancer soul is not required for all routes to the highest goron chest or for all routes to the before boss region. I decided to rework the entire top of the dungeon to account for this.

Water
- The main room Spike enemies can be killed from above with Bow.
- When using Irons to kill the main room spike enemies, you need the Hookshot.
- Zora Tunic shouldn't be hard required for the main room Spike enemies if fewer tunic requirements is enabled.
- [Applies to main branch (except mixed pools isn't implemented yet)] I made a review comment about this one but I guess we all missed that it wasn't addressed. The pots near the boss don't need to check for longshot because they're already in a region that requires longshot, and you want them to not be checking for longshot so that you could get them if you were to come in through the boss door.
- For the enemies under central pillar, we need to confirm adult items to reach the middle water level, since child could have opened it, but child can't do the bottom of the central pillar.
- The before BK shell blade needs to confirm items to get up to the ledge the boulders are coming out of.
- The before BK shell blade can be killed with Din's Fire, without needing to actually swim under to the locked door.
- Stationary helper for the falling platform room keese.
- The before dragon statue tektite was in the region for the chest, which includes the full logic to reach the chest. We don't actually have to solve that dragon statue room. This was made more complicated because only child can swim past the block without a way to dive deeper.
- [Applies to main branch] Added event for making sure you can't get locked out of lowest water level stuff by getting the water stuck at the middle level.

Shadow
- The reverse exit up to the guillotine hallway with longshot was needed to complete the shortcuts reverse loop and kill these enemies potentially without explosives.
- The Wallmaster was in the huge pit region but can't actually be spawned from that region. You'd need the longshot to complete the reverse of the dungeon to get up to it while reversing.
- The Red Bubble here was using the bubble rather than the usual red bubble logic you would see. I decided to require magic for consistency with the other red bubbles.
- You need hookshot to get to the wind tunnel skulltula.
- You need the ReDead and Spinning Pot souls to reverse past the after wind room when dungeon shortcuts are on.

Spirit
- Stationary helper in the first room of child Spirit, for the two stationary keese. (Doesn't actually matter since there's a guaranteed nut drop.)
- Switch to using the new green bubble helper. (Kokiri Sword kinda tricky to get to work due to the short range. Keep targeting the bubble if you want the Stalfos not to attack.)
- We need to complete the reverse loop through child Spirit to reach the checks after the left side bridge without having a means of hitting the switch from afar.
- Stupidly complex logic for the stationary right side keese. They'll all killable without having to go to the far side of the room, using slingshot, boomerang, bombs, or bombchus. Din's Fire can stir them up for a direct kill with sword or sticks. Nuts can stir them up but they almost never fly over to you UNLESS you light the torches first. On the other side of the room, it's kind of annoying to kill them through the gate but it's not too bad if you throw a nut and then immediately jumpslash. I didn't think it was worth having a Stick-only or Sword-only route to kill through the gate. But if you have the rupees then the gate is down and you can lure them over for Sword or Stick without needing nuts or din's or whatever.
- Child Spirit GS had a similar issue with it being placed in the region on the other side of the bridge despite being able to be boomeranged from the near side.
- More precise logic for the Walltulas in child climb. The lowest one on the climb wall can be killed by throwing a pot. (The higher one on the climb wall could technically be killed with a jumpslash but it's a bit too tricky.) The lower of the two by the bomb for light boulder is killable with any jumpslash, though Kokiri Sword is mildly specific to get it to work, and you could potentially lose a few Sticks to it. And the higher of the two by the bomb for light boulder is killable with an adult jumpslash, also mildly specific nut not too bad. All of them were checking for not entrance rando on the bombchu drops route, but that check hasn't been needed for a while now.
- Also fix the Lizalfos logic in here. The current logic could theoretically dead-end you if you spend the key as child and don't find a weapon to fight them with. The current logic is also too restrictive by enforcing the use of 5 keys before ever using any child items. I'm also removing both bombchus and bombs without nuts from logic for Lizalfos.
- The reduced key route to the child side Iron Knuckle wasn't checking for the Beamos, Anubis, and Armos souls.
- The reduced key route for Colossus passthrough wasn't checking for the Beamos, Anubis, and Armos souls.
- The adult side armos can be killed from below. Explosives and Bow are the relevant ways, everything else relevant can be done from the adult climb region. The situation of this check was novel so it was particularly annoying to write the logic for.
- Allowed the walltula in the shifting wall room to be killed without having the necessary items to deal with the beamos for the climb wall, and add additional items that work including hammer, bombchus, and the climb wall trick with bombs. But when using Hook, for consistency with climbing the wall, we need to confirm a way to handle the Beamos.
- Allowed the beamos at the bottom of the shifting wall room without needing to check for something to deal with the walltula first.
- You don't need any additional weapons to kill a white bubble.
- One of the Lizalfos at the top is behind a bombable wall.

Ice
- Using the new helper that allows sticks for individual Freezards.
- The keese in the compass room are stationary, but there's one that's too far from the wall for chus, except that exact one is reasonable with a pot from a previous room.

GTG
- Switch to the new helper for killing the individual Dinolfos.
- Hookshotless GTG trick in the boulder room requires the Wallmaster soul.
- [Applies to main branch] The Wallmaster goes away after the silver rupees have been collected, so the trick to obtain the rupee using the Wallmaster can only be used if silver rupees are not shuffled.
- Hookshotless GTG trick in the Wolfos room requires the Wolfos soul.
- The hookshot target silver rupee in the lava room needs to be in its own region. As part of the hammer room, you might have needed to get it coming from the back of the dungeon while not having the torch slug or keese souls.

Ganon's
- Helper allows Sticks on individual Freezards now.
- [Applies to main branch] Since it's only two Freezards I guess we'll allow Sticks on passing the room as well.
- The Torch Slug in Fire Trial is safe under FTR. There's a trick for child to get the rupee that I guess I'm expanding to include the Torch Slug. But it's not realistic enough to use Bombs to kill it in time (it would have potentially made the trick a lot harder if that could be required).
- One of the bubbles here in Fire Trial is really hard to kill. I decided to allow hovers, with either bombs or magic. To use magic, you have to use great hovers movement to spawn it and then get back onto the platform without having to grab the ledge, and then immediately perform a quick spin. To use bombs, you can set a bomb off the edge, wait a few frames so that you don't spawn the bubble too early, and then use the hover boots to spawn it. But this way the hover boots movement isn't as precise since you can just grab the ledge. Both of these methods require you to perform them twice to do enough damage to kill. There's an easier way that I didn't add to logic because I fear the awkwardness of Biggoron Sword's status as being both a major or minor item depending on context; I didn't feel up to figuring how to correctly allow for Biggoron Sword to be required in logic again, under the enemy drops setting. The Biggoron Sword can do enough damage to kill the bubble in a single slash, and you can perform single slashes while remaining in the hover boots state, which gives you enough time to kill the bubble and obtain the drop before falling into the lava. Master Sword and an adult shield can also kill it in a single hit using a power crouch stab, but I don't think we should ever be calling for power crouch stabs in logic.
- The light trial floormaster comes before the fake wall. No need to put it after the lens/trick check.
- Use the Torch Slug helper in Spirit Trial, which allows for bombs.

Requested Changes:
- Correct typos and misspellings. All remaining instances of Dinalfos should be corrected to Dinolfos. Adjust remaining instances of "Like Like" not being two fully separate words. Adjust capitalization: The first letters of certain phrases or item names were often not capitalized in-game, the D in ReDead was often not capitalized in game and code, the H in Dead Hand was sometimes not capitalized and sometimes there was no space between the words. I'm not sure entirely how many instances of each of these remain.

Some thoughts/suggestions/mild concerns about some of the details of this feature:
- Probably shorten the enemy drops tooltip. (I couldn't figure out how to actually read the whole thing.)
- I feel like we probably have to do something about the two red bubbles that are really difficult to kill, to hopefully guide new players away from shuffling them. I wouldn't agree that not shuffling them at all is the answer, since there are actually ways to kill them and nothing about them is really all that special (other than the difficulty).
- You probably should fix rooms behaving poorly if you don't have an important soul from that room, ie, the Bongo, Morpha, and Barinade boss rooms. I can imagine this could be a bit of a task though...
- You can't currently start with souls using the starting items tab, so it might be worth adding that? (Do you know for sure if they work correctly if a soul goes on a skipped location?)
- Blue Rupee overload? It might be nice if you could implement a variety of original items for enemy drops (and the randomized wonderitem drops as well), perhaps based on the drop tables for those enemies/drops? Or for an easier fix you could just allow the randomized junk pool to take over, just to inject some variety (except for the few enemies that do have a fixed drop item).
- Suggestion to shuffle any enemy drops from the coffins in the Bottom of the Well? (Do the same enemies consistently spawn every time from those?) (Also Suggestion to shuffle Poe Sister item drops? Might be rough since most of them are tied to the paintings.) (Also suggestion to shuffle the Big Octo drop?)
- Probably add Spinning Tile Soul. (Maybe add Poe Soul? Sounds like it could be rough. Paintings again! And I guess there's already an item called "Poe Soul" in this game, so maybe that could be awkward too.)
- Not super important, but it's strange that you have to manually pick up Skull Kid drops. (I haven't heard for sure if you're going to change the way that Skull Kid spawning works. If you do change it, you'll have to remove the logic that I added. If you adjust when they spawn, it should theoretically be the same under shuffled enemy drops as it currently is under adult trade shuffle, as we still want players to be able to trade items in peace.)
- I'm not quite happy with the way that you obtain drops from Deku Babas. I feel like you should also just get the randomized item whenever you would spawn Deku Nuts?
- Enemies that spawn more enemies seem especially weird about how you get the drops. I've heard you might have to kill a Biri to get the drop from Baris, but I've never actually this, so I'd like to see some proof that it can actually occur lol. (It might have logic impact if it is actually to be required to kill any of the Biris that spawn since you might have to take damage multiple times to kill them all.) I've heard Floormaster might give the drop when you kill the main one, but I've never seen this either. Amoung the small ones, I'm not sure how it decides which has the drop. (If the main one is what will have the drop in the future, it might require a small but annoying logic change since bombs are fine without nuts on the big Floormaster.) But the Peahats definitely sometimes give the drops at strange times, including potentially before you've actually done enough damage to the Peahat to kill it? I have no idea what I did to cause this but I've seen it a couple of times now. More common is that you get the drop after you damage it to death but before the animation finishes and drop should actually spawn.
- I think the soul region for grotto enemies to spawn in regional souls should depend on the parent region of the grotto; no separate grotto soul to find. Treating grottos as their own region contradicts with how the rest of randomizer treats grottos. Also, graves are usually considered Grottos in randomizer, so if a Grotto soul exists one might expect the enemies in graves to obey the Grotto soul rather than the Graveyard soul.
- The enemies on the minimap feature has some slight issues, with the dots in areas that don't have maps typically appearing in the middle of the screen instead of where the map would go. There was also some slightly strange dot placements as you change between rooms. I also found the tiny 1px dots really hard to see on my TV -- I'm on VC using a mildly snowy CRT. Especially against the blue of a minimap, the dots showed up only as a slightly purple-tinged pixel and it felt a bit like playing Where's Waldo when trying to spot them. Maybe making them a 2px square could probably go a long way to making them more visible? It does say "Beta" though, which is basically a get out of jail free card.
- And of course we've discussed this one before, but I don't agree that the flying pot soul should be a requirement for clearing rooms with flying pots, since you don't actually need to defeat them unless you stir them up (and if they're not there you can't stir them up).

fenhl commented:
> * [Applies to main branch] Corrected the wonderitem vanilla drops. (I decided to go with Blue Rupee for wonderitems that normally give random drops.)

I suggest using new items, named “Drop Table N” or something, which can be aliased to blue rupees for now (same get item ID and logically aliased, see `Piece of Heart (Treasure Chest Game)` for an existing example), to make it easier to change the behavior of these items later. For example, you could imagine having the item pool code replace them with one of the possible drops, or actually adding new get item IDs that matches the vanilla behavior.

r0bd0g commented:
I did leave comments in the location list about which wonderitem drops were randomized so that they'll be easy to find later (it's all of the Hideout Bow wonderitems [not the stove], and one Hookshot wonderitem in MQ Water).

And later in the thoughts/suggestions section of my PR comments I basically suggest implementing drop tables. Especially now with enemy drop shuffle I was feeling a bit inundated with blue rupees.

(Now that we have a magic meter item over on main branch though, we might have to go back through these drops again and add magic jars once we finally add drops for those. Honestly I should have commented about which locations have those as well. There was a least one drop that seemed to be a small magic as child and a large magic as adult, I think? I used 20 Rupees in place of all wonderitem magic drops [enemies that only drop magic weren't touched], which I think was the intention before.)

r0bd0g commented:
This branch split off from main branch somewhere before Pots 3.0 was complete. Especially the location list file I mangled kind of all over the place with this PR, but there are some flag changes and things from Pots 3.0 that still need to brought in. So it might be a bit annoying to update to main without losing accidentally introducing some regressions (I noticed at least some flags changing on child kak crates, and the shadow v/mq spinning pots). Some other logic-file impacting PRs that have come through since then include the big poe misc hint, and child fortress poh shuffle. (I could see 100gs shuffle likely making it in soon as well, not sure what else might be in the pipe.)

r0bd0g commented:
The change that I made to skull kid logic is apparently not necessary, since I didn't realize the change to skull kid spawning applied to all settings. So if the skull kids don't spawn you can just show all relevant trade items and then they will. So my logic change should be removed. I added an event for them that you can just delete, and then revert skull kid logic to adult+soul, as it was before. Do you need me to make that change here?

r0bd0g commented:
I removed the (undocumented -- forgot to write it into this PR description) fix to skull kids spawning b/c I incorrectly assumed the change was only made under adult trade shuffle.

Fixed long-standing main branch issue with the mq ice sot gs not checking for bfas as an alternative to sot. I think this might be the only case of bfa-specific logic anywhere in rando?

rrealmuto commented:
This PR has been merged. I rebased it locally as #53bd19d1f25d5ad9651e4de28dfb8c6b7117591c, so github won't see that. Therefore I am closing this PR. I'll add an issue capturing your requested changes/additional thoughts.

r0bd0g commented:
I guess for the record, I had a bug with the bubble under the fire temple bridge failing to check for longshot, that you caught and fixed. There was a fix to skull kid that I forgot to document in the text that isn't actually needed, but I was made aware of this too late to get the change into this PR. And the documented main branch fix to MQ Ice SoT skull also came late and didn't make it in. I'll come back for these next review if they're still there.
